parquet-variant/src/main/java/org/apache/parquet/variant/VariantUtil.java

```diff
@@ -0,0 +1,843 @@
+    /**
+     * @param arrayInfo The details of the Variant array
+     */
+    T apply(ArrayInfo arrayInfo) throws IOException;
+  }
+
+  /**
+   * A helper function to access a Variant array, at `value[pos...]`.
+   * @param value The Variant value
+   * @param pos The starting index of the Variant value
+   * @param handler The handler to process the array
+   * @return The result of the handler
+   * @param <T> The return type of the handler
+   */
+  public static <T> T handleArray(byte[] value, int pos, ArrayHandler<T> handler) {
+    ArrayInfo info = parseArray(value, pos);
+    return handler.apply(info);
+  }
+
+  /**
+   * Same as `handleArray` but handler can throw IOException.
+   */
+  public static <T> T handleArrayException(byte[] value, int pos, ArrayHandlerException<T> handler)
+      throws IOException {
+    ArrayInfo info = parseArray(value, pos);
+    return handler.apply(info);
+  }
+
+  /**
+   * Parses the array at `value[pos...]`, and returns the array details.
+   */
+  private static ArrayInfo parseArray(byte[] value, int pos) {
+    checkIndex(pos, value.length);
+    int basicType = value[pos] & BASIC_TYPE_MASK;
+    int typeInfo = (value[pos] >> BASIC_TYPE_BITS) & PRIMITIVE_TYPE_MASK;
+    if (basicType != ARRAY) {
+      throw unexpectedType(Type.ARRAY);
+    }
+    // Refer to the comment of the `ARRAY` constant for the details of the object header encoding.
+    // Suppose `typeInfo` has a bit representation of 000_b2_b1b0, the following line extracts
+    // b2 to determine whether the object uses a 1/4-byte size.
+    boolean largeSize = ((typeInfo >> 2) & 0x1) != 0;
+    int sizeBytes = (largeSize ? U32_SIZE : 1);
+    int numElements = readUnsigned(value, pos + 1, sizeBytes);
+    // Extracts b1b0 to determine the integer size of the offset list.
+    int offsetSize = (typeInfo & 0x3) + 1;
+    int offsetStart = pos + 1 + sizeBytes;
+    int dataStart = offsetStart + (numElements + 1) * offsetSize;
+    return new ArrayInfo(numElements, offsetSize, offsetStart, dataStart);
+  }
+
+  /**
+   * Returns a key at `id` in the Variant metadata.
+   * @param metadata The Variant metadata
+   * @param id The key id
+   * @return The key
+   * @throws MalformedVariantException if the Variant is malformed or if the id is out of bounds
+   */
+  public static String getMetadataKey(byte[] metadata, int id) {
+    checkIndex(0, metadata.length);
+    // Extracts the highest 2 bits in the metadata header to determine the integer size of the
+    // offset list.
+    int offsetSize = ((metadata[0] >> 6) & 0x3) + 1;
+    int dictSize = readUnsigned(metadata, 1, offsetSize);
+    if (id >= dictSize) {
+      throw new MalformedVariantException(
+          String.format("Invalid dictionary id: %d. dictionary size: %d", id, dictSize));
+    }
+    // There are a header byte, a `dictSize` with `offsetSize` bytes, and `(dictSize + 1)` offsets
+    // before the string data.
+    int stringStart = 1 + (dictSize + 2) * offsetSize;
```

Comment:
I think it would be easier to read this if you also used `offsetListOffset` to capture `1 + dictSize`.